toolkit/tools/pkg/imagecustomizerlib/defaultgrubutils.go

```diff
@@ -0,0 +1,256 @@
+	defaultGrubCmdlineLinuxDefault = "GRUB_CMDLINE_LINUX_DEFAULT"
+	grubKernelOpts                 = "kernelopts"
+
+	// The variable in the /etc/default/grub file that contains the SELinux args.
+	defaultGrubVarCmdlineForSELinux = defaultGrubCmdlineLinux
+)
+
+type grubDefaultVarAssign struct {
+	Token grub.Token
+	Name  string
+	Value string
+}
+
+func findGrubDefaultVarAssigns(grubDefaultConfig string) ([]grubDefaultVarAssign, error) {
+	// So, technically the /etc/default/grub file is a Bash file not a grub config file.
+	// While these are very similar formats, they do have differences.
+	// But, we are just going to hope that the user isn't annoying enough to Bash specifc features within
+	// /etc/default/grub file. (e.g. <<< strings.)
+	grubTokens, err := grub.TokenizeConfig(grubDefaultConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	varAssigns := []grubDefaultVarAssign(nil)
+
+	lines := grub.SplitTokensIntoLines(grubTokens)
+	for _, line := range lines {
+		if len(line) != 1 {
+			// Normal variable assignments only have 1 value.
+			// Export variable assignments have 2 values. But we are ignoring those.
+			continue
+		}
+
+		argToken := line[0]
+
+		isVarAssign := len(argToken.SubWords) >= 1 &&
+			argToken.SubWords[0].Type == grub.KEYWORD_STRING &&
+			strings.Contains(argToken.SubWords[0].Value, "=")
+
+		if !isVarAssign {
+			continue
+		}
+
+		argStringBuilder := strings.Builder{}
+		for _, subword := range argToken.SubWords {
+			switch subword.Type {
+			case grub.KEYWORD_STRING, grub.STRING:
+				argStringBuilder.WriteString(subword.Value)
+
+			case grub.QUOTED_VAR_EXPANSION:
+				if subword.Value != grubKernelOpts {
+					return nil, fmt.Errorf("unexpected quoted variable expansion (%s)", subword.Value)
+				}
+
+				// There is a bug in Azure Linux 2.0 and early builds of Azure Linux 3.0 where "$kernelopts" is not
+				// properly escaped.
+				// So, be nice and fix it up.
+				argStringBuilder.WriteString("$")
+				argStringBuilder.WriteString(grubKernelOpts)
+
+			case grub.VAR_EXPANSION:
+				return nil, fmt.Errorf("unexpected variable expansion (%s)", subword.Value)
+			}
+		}
+
+		argValue := argStringBuilder.String()
+		name, value, foundEqSymbol := strings.Cut(argValue, "=")
+		if !foundEqSymbol {
+			// Not a variable assignment.
+			continue
+		}
+
+		varAssign := grubDefaultVarAssign{
+			Token: argToken,
+			Name:  name,
+			Value: value,
+		}
+		varAssigns = append(varAssigns, varAssign)
+	}
+
+	return varAssigns, nil
+}
+
+func findGrubDefaultVarAssign(varAssigns []grubDefaultVarAssign, name string) (grubDefaultVarAssign, error) {
+	for _, varAssign := range varAssigns {
+		if varAssign.Name == name {
+			return varAssign, nil
+		}
+	}
+
+	err := fmt.Errorf("failed to find %s variable assignment (%s)", installutils.GrubDefFile, name)
+	return grubDefaultVarAssign{}, err
+}
+
+func getGrubDefaultLinuxArgs(defaultGrubContent string, varName string,
+) (grubDefaultVarAssign, []grubConfigLinuxArg, int, error) {
```

Comment:
What would grubDefaultVarAssign contain?
The entire: `GRUB_CMDLINE_LINUX_DEFAULT=...` line?

Reply:
If the function is to return/find the linux arguments (from the name), why does it take a varName?
I think there's ambiguity re. varName - is this `GRUB_CMDLINE_LINUX_DEFAULT` or a kernel command line argument argument?

Reply:
Alas, there are two variables in the `/etc/default/grub` file that contains command line args: `GRUB_CMDLINE_LINUX_DEFAULT` and `GRUB_CMDLINE_LINUX`. `varName` specifies which variable to look at.

Reply:
Yes, the `grubDefaultVarAssign` value contains details for the `GRUB_CMDLINE_LINUX_DEFAULT=...` line.

The `grubDefaultVarAssign.Token` field contains the start and end locations of the line in the `/etc/default/grub` file.